src/tls/main.c

```diff
@@ -0,0 +1,146 @@
+struct arguments {
+  uint16_t port;
+  int no_tls;
+  unsigned idle_timeout;
+};
+
+#define OPT_NO_TLS 1000
+#define OPT_IDLE_TIMEOUT 1001
+
+static unsigned
+arg_parse_unsigned (char *arg, struct argp_state *state, unsigned min, unsigned max, const char *error_msg)
+{
+  char *endptr = NULL;
+  long num = strtol (arg, &endptr, 10);
+
+  if (!*arg || *endptr != '\0' || num < min || num > max)
+    argp_error (state, "%s: %s", error_msg, arg);
+  return (unsigned) num;
+}
+
+/* Parse a single option. */
+static error_t
+parse_opt (int key, char *arg, struct argp_state *state)
+{
+  struct arguments *arguments = state->input;
+
+  switch (key)
+    {
+      case OPT_NO_TLS:
+        arguments->no_tls = 1;
+        break;
+      case 'p':
+        arguments->port = arg_parse_unsigned (arg, state, 1, UINT16_MAX, "Invalid port");
+        break;
+      case OPT_IDLE_TIMEOUT:
+        arguments->idle_timeout = arg_parse_unsigned (arg, state, 0, (UINT_MAX / 1000) - 1, "Invalid idle timeout") * 1000;
+        break;
+      default:
+        return ARGP_ERR_UNKNOWN;
+    }
+  return 0;
+}
+
+static struct argp_option options[] = {
+  {"no-tls", OPT_NO_TLS, 0, 0,  "Don't use TLS" },
+  {"port", 'p', "PORT", 0, "Local port to bind to (9090 if unset)" },
+  {"idle-timeout", OPT_IDLE_TIMEOUT, "SECONDS", 0, "Time after which to exit if there are no connections; 0 to run forever (default: 90)" },
+  { 0 }
+};
+
+static const struct argp argp = {
+  .options = options,
+  .parser = parse_opt,
+  .doc = "cockpit-tls -- TLS terminating proxy for cockpit-ws",
+};
+
+Server *server;  /* global, can't pass it to our signal handler */
+
+static void
+handle_sigchld (int signal, siginfo_t *info, void *ucontext)
+{
+  debug ("got SIGCHLD: pid %u exited with status %x", info->si_pid, info->si_status);
+  server_remove_ws (server, info->si_pid);
+}
+
+int
+main (int argc, char **argv)
+{
+  struct arguments arguments;
+  const struct sigaction child_action = {
+    .sa_sigaction = handle_sigchld,
+    .sa_flags = SA_SIGINFO | SA_NOCLDSTOP
+  };
+  char *error = NULL;
+  char *certfile = NULL;
+
+  /* default option values */
+  arguments.no_tls = 0;
+  arguments.port = 9090;
+  arguments.idle_timeout = 90000;
+
+  argp_parse (&argp, argc, argv, 0, 0, &arguments);
+
+  if (!arguments.no_tls)
+    {
+      certfile = cockpit_certificate_locate (&error);
+      if (error)
+        {
+          warnx ("Could not locate server certificate, disabling TLS: %s", error);
+          free (error);
+        }
+      debug ("Using certificate %s", certfile);
+    }
+
+  /* TODO: Add cockpit.conf option to enable client-certificate auth, once we support that */
+  server = server_new (COCKPIT_WS, arguments.port, certfile, NULL, CERT_NONE, arguments.idle_timeout);
+  free (certfile);
+
+  /* track cockpit-ws children */
+  if (sigaction (SIGCHLD, &child_action, NULL) < 0)
+      err (1, "failed to set up SIGCHLD handler");
+
+  while (server_poll_event (server, 1000))
```

Comment:
If this `1000` is what I think it is, then I'm going to be very upset.

separately: why not just have a server_run() method or something?

Reply:
Thanks for pointing out, what was I thinking.. I cleaned this up, it's now both simpler and also avoids the non-busy wakeups.